src/burnchains/bitcoin/blocks.rs

```diff
@@ -307,6 +308,112 @@ impl BitcoinBlockParser {
         }
     }
 
+    /// Parse op_drop data from the first input to get a byte payload
+    ///
+    /// The data may be present in the redeem script directly or in the witness
+    ///
+    /// Note: We initially only support p2sh, p2tr-over-p2sh and p2wsh-over-p2sh.
+    ///       Supporting plain p2wsh and p2tr require having the scriptPubKey available,
+    ///       which entails a UTXO lookup and require substantial codebase changes to implement.
+    fn parse_op_drop_data(input: &TxIn) -> Option<(u8, Vec<u8>)> {
+        let instructions = bits::parse_script(&input.script_sig);
+
+        // TODO(3680): Test this logic and add appropriate logging
+        match instructions.as_slice() {
+            &[] => {
+                test_debug!("Empty script sig suggests native segwit which is not supported.");
+                None
+            }
+            &[Instruction::PushBytes(redeem_script)] => {
+                test_debug!(
+                    "Parsing OP_DROP data from redeem script: {:?}",
+                    redeem_script
+                );
+                Self::parse_op_drop_data_from_redeem_script(
+                    &Script::from(redeem_script.to_vec()),
+                    &input.witness,
+                )
+            }
+            _ => {
+                test_debug!("Invalid script sig format");
+                None
+            }
+        }
+    }
+
+    fn parse_op_drop_data_from_redeem_script(
+        redeem_script: &Script,
+        witness: &[impl AsRef<[u8]>],
+    ) -> Option<(u8, Vec<u8>)> {
+        let instructions = bits::parse_script(redeem_script);
+
+        match instructions.as_slice() {
+            // Segwit v0
+            &[Instruction::PushBytes([]), Instruction::PushBytes(witness_hash)]
+                if witness_hash.len() == 32 =>
+            {
+                test_debug!("Parsing segwit v0 data");
+                Self::parse_op_drop_data_from_segwit_v0_witness(witness)
+            }
+            // Taproot
+            &[Instruction::Op(btc_opcodes::OP_PUSHNUM_1), Instruction::PushBytes(witness_hash)]
+                if witness_hash.len() == 32 =>
+            {
+                test_debug!("Parsing taproot data");
+                Self::parse_op_drop_data_from_taproot_witness(witness)
+            }
```

Comment:
p2tr-over-p2sh is not a thing. I have to change this PR to support native p2tr OP_DROP data.